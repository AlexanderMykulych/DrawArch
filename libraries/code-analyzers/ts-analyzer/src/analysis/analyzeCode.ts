import { progress } from "../progress/progress";
import { createContextForOneFile } from "./context/analysisContext";
import { getAnalysisPlugins } from "./plugins/getAnalysisPlugins";
import { preparePlugins } from "./preparePlugins";
import type { CodeInfo } from "./types";

export type AnalyzeCodeData = {
  code: string
  fileName: string
}

async function _analyzeCode({ code, fileName }: AnalyzeCodeData): Promise<CodeInfo[]> {
  const context = await createContextForOneFile({
    code,
    fileName,
  })

  const plugins = getAnalysisPlugins()
  const preparedPluginsWithContext = await preparePlugins(plugins, context)

  const initTasks = preparedPluginsWithContext
    .map(({ plugin, context }) => plugin.init && plugin.init(context))

  await Promise.all(initTasks)

  const runTasks = preparedPluginsWithContext
    .map(({ plugin, context }) => plugin.run(context))
  
  await Promise.all(runTasks)

  return context.result
}

export const analyzeCode = progress('analyzeCode', _analyzeCode)