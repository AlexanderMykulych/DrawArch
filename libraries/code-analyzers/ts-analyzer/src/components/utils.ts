import { Node } from "ts-morph";

export function getNodeName(node: Node | undefined): string | undefined {
  if (Node.isVariableDeclaration(node)) {
    return node.getName()
  }
  if (Node.isFunctionDeclaration(node)) {
    return node.getName()
  }
  if (Node.isParameterDeclaration(node)) {
    return node.getName()
  }
  if (Node.isExportSpecifier(node)) {
    return node.getName()
  }
}