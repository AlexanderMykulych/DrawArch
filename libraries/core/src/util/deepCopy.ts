export function deepCopy(val: any) {
    return !!val ? JSON.parse(JSON.stringify(val)) : val;
}