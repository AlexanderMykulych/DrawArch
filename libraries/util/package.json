{
  "version": "0.9.20",
  "license": "MIT",
  "main": "dist/coya-util.es.js",
  "typings": "dist/src/index.d.ts",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "start": "tsdx watch",
    "build": "vite build",
    "test": "tsdx test",
    "lint": "tsdx lint",
    "prepare": "tsdx build",
    "size": "size-limit",
    "analyze": "size-limit --why"
  },
  "husky": {
    "hooks": {
      "pre-commit": "tsdx lint"
    }
  },
  "prettier": {
    "printWidth": 80,
    "semi": true,
    "singleQuote": true,
    "trailingComma": "es5"
  },
  "name": "coya-util",
  "author": "AlexMykulych",
  "module": "dist/coya-util.es.js",
  "size-limit": [
    {
      "path": "dist/coya-util.cjs.production.min.js",
      "limit": "10 KB"
    },
    {
      "path": "dist/coya-util.esm.js",
      "limit": "10 KB"
    }
  ],
  "devDependencies": {
    "@size-limit/preset-small-lib": "^7.0.5",
    "husky": "^7.0.4",
    "size-limit": "^7.0.5",
    "tsdx": "^0.14.1",
    "tslib": "^2.3.1",
    "typescript": "^4.5.5"
  },
  "dependencies": {
    "@rollup/plugin-typescript": "^8.3.0",
    "curved-arrows": "^0.1.0",
    "hex-to-rgba": "^2.0.1",
    "rollup-plugin-typescript2": "^0.31.1",
    "svg-path-commander": "^0.1.25",
    "vite": "^2.8.0",
    "vite-plugin-dts": "~1.2.1"
  }
}
