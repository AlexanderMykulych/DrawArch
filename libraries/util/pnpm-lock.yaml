lockfileVersion: 5.3

specifiers:
  fast-deep-equal: ^3.1.3

dependencies:
  fast-deep-equal: 3.1.3

packages:

  /fast-deep-equal/3.1.3:
    resolution: {integrity: sha512-f3qQ9oQy9j2AhBe/H9VC91wLmKBCCU/gDOnKNAYG5hswO7BLKj09Hc5HYNz9cGI++xlpDCIgDaitVs03ATR84Q==}
    dev: false
