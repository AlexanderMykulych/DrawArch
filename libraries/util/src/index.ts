import SVGPathCommander from 'svg-path-commander';
export * from './setValueByPath';
export * from './whatChanged';
export * from './deepAssign';
export * from './deepCopy';
export * from './fastDeepEqual';
export * from './isNotNullOrUndefined';
export * from './isNullOrUndefined';
export * from './hexToRgba';
export * from './groupBy';
export const svgPathCommander = SVGPathCommander;
