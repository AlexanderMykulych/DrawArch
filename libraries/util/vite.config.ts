import { defineConfig } from 'vite'
import path from 'path';
// import typescript from '@rollup/plugin-typescript';
import typescript from 'rollup-plugin-typescript2'
const resolvePath = (str: string) => path.resolve(__dirname, str)
// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        // typescript({
        //     'target': 'es2020',
        //     'rootDir': resolvePath('./src'),
        //     'declaration': true,
        //     'declarationDir': resolvePath('./dist'),
        //     exclude: resolvePath('./node_modules/**'),
        //     allowSyntheticDefaultImports: true
        // })
        typescript({
            check: false,
            tsconfig: resolvePath('./tsconfig.json'),
            tsconfigOverride: {
                compilerOptions: {
                    sourceMap: false,
                    declaration: true,
                    declarationMap: false,
                    target: 'es2020',
                },
                exclude: ['**/__tests__']
            }
        }),
    ],
    build: {
        rollupOptions: {
            output: {
                format: 'es'
            },
        },
        sourcemap: true,
        lib: {
            entry: resolvePath("/src/index.ts"),
            name: "coya-util",
            fileName: (format) => `coya-util.${format}.js`,
            formats: ['es']
        },
    }
})
