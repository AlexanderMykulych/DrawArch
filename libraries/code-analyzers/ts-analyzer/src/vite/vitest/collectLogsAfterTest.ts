import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}