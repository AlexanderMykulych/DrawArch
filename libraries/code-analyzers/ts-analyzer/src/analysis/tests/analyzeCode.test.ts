import { test, expect } from 'vitest'
import { analyzeCode } from '../analyzeCode'

test('should analyze code as string', async () => {
  const code = `
    function f1() {
      return 1
    }
  `
  const fileName = 'funcs.ts'

  const codeInfos = await analyzeCode({
    code,
    fileName,
  })

  expect(codeInfos).toMatchSnapshot()
  expect(codeInfos.length).greaterThan(0)
})

test.only('should analyze vue code as string', async () => {
  const code = `
<script lang="ts">
import { fn1 } from './utils/util';

export default {
  data() {
    return {
      val: fn1(1)
    }
  }
}
</script>

<template>
  <div>
    {{val}}
  </div>
</template>
  `
  const fileName = 'cmp.vue'

  const codeInfos = await analyzeCode({
    code,
    fileName,
  })

  expect(codeInfos.length).greaterThan(0)
})
