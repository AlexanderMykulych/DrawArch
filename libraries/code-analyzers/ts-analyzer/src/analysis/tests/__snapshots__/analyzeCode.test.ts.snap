// Vitest Snapshot v1

exports[`should analyze vue code as string 1`] = `
[
  {
    "from": "/cmp.vue",
    "relationType": "import",
    "to": "./utils/util",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/cmp.vue",
    "id": "/cmp.vue",
    "type": "entity",
  },
  {
    "end": 36,
    "entityType": "import_declaration",
    "filePath": "<unknow>",
    "id": "./utils/util",
    "kind": "ImportDeclaration",
    "lineEnd": 2,
    "lineStart": 2,
    "start": 1,
    "type": "entity",
  },
  {
    "from": "/cmp.vue",
    "relationType": "import",
    "to": "vue",
    "type": "relationship",
  },
  {
    "end": 235,
    "entityType": "import_declaration",
    "filePath": "<unknow>",
    "id": "vue",
    "kind": "ImportDeclaration",
    "lineEnd": 12,
    "lineStart": 12,
    "start": 110,
    "type": "entity",
  },
  {
    "code": " {
  return (_openBlock(), _createElementBlock(\\"div\\", null, _toDisplayString(_ctx.val), 1 /* TEXT */))
}",
    "end": 377,
    "entityType": "function",
    "filePath": "/cmp.vue",
    "id": "/cmp.vue/render",
    "kind": "FunctionDeclaration",
    "lineEnd": 16,
    "lineStart": 14,
    "logStart": 274,
    "start": 237,
    "type": "entity",
    "typeString": "typeof import(\\"/cmp.vue\\").render",
  },
  {
    "end": 179,
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue/openBlock",
    "kind": "ImportSpecifier",
    "lineEnd": 12,
    "lineStart": 12,
    "start": 156,
    "type": "entity",
  },
  {
    "end": 222,
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue/createElementBlock",
    "kind": "ImportSpecifier",
    "lineEnd": 12,
    "lineStart": 12,
    "start": 181,
    "type": "entity",
  },
  {
    "end": 154,
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue/toDisplayString",
    "kind": "ImportSpecifier",
    "lineEnd": 12,
    "lineStart": 12,
    "start": 119,
    "type": "entity",
  },
  {
    "from": "/cmp.vue/render",
    "relationType": "use",
    "to": "/cmp.vue/openBlock",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue/render",
    "relationType": "use",
    "to": "/cmp.vue/createElementBlock",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue/render",
    "relationType": "use",
    "to": "/cmp.vue/toDisplayString",
    "type": "relationship",
  },
  {
    "code": " {
    return {
      val: fn1(1)
    }
  }",
    "end": 106,
    "entityType": "function",
    "filePath": "/cmp.vue",
    "id": "/cmp.vue/data",
    "kind": "MethodDeclaration",
    "lineEnd": 9,
    "lineStart": 5,
    "logStart": 64,
    "start": 57,
    "type": "entity",
    "typeString": "() => { val: any; }",
  },
  {
    "end": 13,
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue/fn1",
    "kind": "ImportSpecifier",
    "lineEnd": 2,
    "lineStart": 2,
    "start": 10,
    "type": "entity",
  },
  {
    "from": "/cmp.vue/data",
    "relationType": "use",
    "to": "/cmp.vue/fn1",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue",
    "relationType": "parent",
    "to": "/cmp.vue/render",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue",
    "relationType": "parent",
    "to": "/cmp.vue/data",
    "type": "relationship",
  },
]
`;
